Define entity attributes as entity class variables

frenck opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This is a proposal to add class variables to our entity base classes.

In this PR, I've limited the scope to the `Entity` class.

These entity attributes are properties that are implemented statically most of the time, adding quite a bit of boilerplate code across our codebase.

So, instead of

```py
    @property
    def device_class(self) -> str:
        """Return the class of this entity."""
        return DEVICE_CLASS_TEMPERATURE
```

We can also (besides this one above) accept the following in the constructor:

```py
    self._attr_device_class = DEVICE_CLASS_TEMPERATURE
```

Or even as class variable itself

```py
class CoronavirusSensor(CoordinatorEntity, SensorEntity):
    """Sensor representing corona virus data."""

    _attr_unit_of_measurement = "people"
```

In all the above cases, types would still be checked. I tried this PR in the past (which didn't work). With recent MyPy changes/fixes, this now became an option.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

frenck committed:
Define entity attributes as entity class variables

frenck committed:
Example coronavirus integration

frenck committed:
Example verisure

frenck committed:
Cleanup/typing fixes

frenck committed:
Fix Coronavirus

MartinHjelmare commented:
In general I'm in favor.

A risk when adding this may be that private attributes that already exist with the here chosen names will be set as device class, unit and icon while before that wasn't done. So the wrong things may be set  unintentionally for our interface.

frenck commented:
> A risk when adding this may be that private attributes that already exist with the here chosen names will be set as device class, unit and icon while before that wasn't done.

Yeah, that was my first scare as well. So, that should not be an issue:

- Property methods are currently overwritten everywhere, thus the private class variable isn't used.
- If the typing of the existing class variable isn't there, mypy would fail (which is the reason for 659b79e)

Considering that, I think we could implement this fairly safely and gradually.

Alternatively, we could also choose to prefix them, e.g., using `attr_unit_of_measurement` instead of `_unit_of_measurement`.

MartinHjelmare commented:
A risk case could be an entity that doesn't implement eg the `device_class` property but has a private attribute `_device_class` with a string value that represents something else than an `Entity.device_class`. We wouldn't catch that, right?

frenck commented:
That would indeed not be caught. I assume the risk for `_device_class` is fairly minimal, however, for other properties implementing this structure it might be.

For example, `_name`.

So that means we either accept the risk or make it less risky by prefixing it as suggested in the previous post (but that is uglier as well IMHO).

---

3rd alternative might be non-Final marked constants.

```
class CoronavirusSensor(CoordinatorEntity, SensorEntity):
    """Sensor representing corona virus data."""

    UNIT_OF_MEASUREMENT = DEVICE_CLASS_TEMPERATURE
```

Although this looks a bit weird as well, it does make clear value should be constant (otherwise use the property)

KapJI commented:
I like that 3rd alternative, it makes it more consistent with config flow constants (`VERSION`). It this also makes it more clear that this is something what is consumed externally. When setting a private field not using `super()` constructor for the parent class is a bit weird.

frenck commented:
@KapJI `VERSION` isn't declared... 😉  But for that case it works.

frenck commented:
OK, so a bit of research done on this variant

```py
    def __init__(self, coordinator, country, info_type):
        """Initialize coronavirus sensor."""
        self.DEVICE_CLASS = DEVICE_CLASS_TEMPERATURE
```

Which triggers Pylint, C0103. Pylint considers everything accessed with `self.` an attribute, not a class attribute. This is weird, as defining a class attribute has a different regex rule. Thus you can define class attributes that have a valid name, while accessing them isn't allowed 🤷

MartinHjelmare commented:
Maybe we can fix that?

frenck commented:
Looking at the issue tracker of Pylint, it is deliberate, they can be used by class methods.

As for our use-case, if we want to go the "constants" route, we can do either of 2 things:
Adjust the attribute regex OR add known attributes as `good-names` to Pylint's configuration.

bdraco commented:
I like the `good-names` solution as it keeps the scope limited to this case

bdraco commented:
`nut` is another good test case https://github.com/home-assistant/core/blob/dev/homeassistant/components/nut/sensor.py#L80

frenck commented:
Alternative implementation in #50931

balloob commented:
This would be the ideal case from my perspective:

```python
class Entity:

    icon: str | None = None


class MyDynamicEntity(Entity):

	@property
    def icon(self) -> str | None:
        return ICONS.get(self._device.type)


class MyStaticEntity(Entity):

	def __init__(self) -> None:
        self.icon = ICONS.get(self._device.type)

```

frenck commented:
@balloob Yes, but that isn't viable for both mypy & pylint.

balloob commented:
Ok I retract my idea. Although we could maybe make it work with MyPy, it would only be a matter of time until we want to try something else (like Instagram's Static Python) that consumes types and chokes on this.

balloob commented:
I think that this PR is my preferred approach. I don't think we will see any issues where people use one of these attributes and don't have a property defined to return it.

frenck committed:
Define entity attributes as entity class variables (attr alternative)

frenck committed:
Example coronavirus

frenck committed:
Example nut

frenck committed:
Example verisure

frenck commented:
Ok final alternative: 

In: https://github.com/home-assistant/core/pull/50933

balloob commented:
I am 👍 on @KapJI reasoning.

frenck committed:
Mark private

frenck commented:
Yeah so, I did make a change to the `attr` variant.

As the class attributes are not meant to be consumed outside of the class, I've marked them private.

Which results in `_attr`

frenck committed:
Revert "Fix Coronavirus"

This reverts commit 060843860fe300f8448d0d2476de8963d5ddf5a2.

frenck committed:
Revert "Cleanup/typing fixes"

This reverts commit 659b79e75a97007f7181064e446c3e988c2d54bb.

frenck committed:
Cleanup after all reverting/cherrypicking/merging

frenck commented:
Applied the alternative onto this PR (to keep the code change and discussions together)

frenck committed:
Implement all entity properties

frenck committed:
Update coronavirus example

frenck committed:
Update nut example

frenck committed:
Update verisure example

frenck committed:
Lets not talk about this one...

frenck commented:
Alright, updated processed the discussions/variants, completed the transition for the Entity helper and updated the 3 integrations used as an example.

Additionally opened up a PR for the developer documentation.

MartinHjelmare commented on homeassistant/components/coronavirus/sensor.py:
```diff
@@ -27,17 +27,21 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
-    name = None
-    unique_id = None
+    _attr_unit_of_measurement = "people"
+    _attr_extra_state_attributes = {ATTR_ATTRIBUTION: ATTRIBUTION}
```
It's a bit dangerous with mutable class attributes. If it's modified all instances will inherit the change.

frenck replied:
Thanks! Adjusted.

frenck committed:
Fix multiple class attribute

rytilahti commented:
I really like the fact that this allows getting rid of the copy&pasting the boilerplate code, but as a downside, this will make it even harder to follow the code flow with the dynamic "magic" (either when just trying to grep for the code paths or using a debugger).

Just a thought and not something to be considered as a proper proposal, but is there something against having platform specific "metadata"/"info"/"state" classes?

A very quickly drafted example:
1. `LigthtEntity` implementations are required to implement `state() -> LightState`
2. `LightState` is defined to extend `State`, containing the information that light entities can/should expose, such as as the active color temperature or brightness.
3. `State` defines a container for common attributes that all entities MUST define (e.g., state, name, unit of measurement, ..)
4. Similarly, a separate `Metadata` for non-changing attributes could be introduced if it's unwanted to put this into the `State`. For example, `LightEntity` would have an attribute to mark the (optional) color temperature range etc.

Some thoughts about this approach vs. magic class variables:
* This would allow having a common, platform defined API that MUST be implemented. This can then be enforced by static checks to avoid pitfalls for new developers forgetting to define some attribute.
* There is nothing magical about it, as a developer you can clearly see what you need to implement. For runtime compatibility, the base class implementations can inform when some attribute is not defined (e.g, when the API changes in the future).
* If and when the API changes, unimplemented attributes can be reported to the developer/users using the logger. On other cases, linters will inform that the state class is not conform.
* The platforms / main entity class accessors (e.g., `icon()`, `name()`, etc.) keep working as they are, just as

frenck commented:
> State defines a container for common attributes that all entities MUST define (e.g., state, name, unit of measurement, ..)

That is what our entity classes are?
